pkg/snmp/traps/formatter.go

```diff
@@ -0,0 +1,249 @@
+		data["snmpTrapMIB"] = trapMetadata.MIBName
+	}
+
+	parsedVariables := parseVariables(variables[2:])
+	data["variables_raw"] = parsedVariables
+	for _, variable := range parsedVariables {
+		varMetadata, err := f.oidResolver.GetVariableMetadata(trapOID, variable.OID)
+		if err != nil {
+			log.Debugf("unable to enrich variable: %s", err)
+			continue
+		}
+		data[varMetadata.Name] = variable.Value
+	}
+	return data, nil
+}
+
+// NormalizeOID convert an OID from the absolute form ".1.2.3..." to a relative form "1.2.3..."
+func NormalizeOID(value string) string {
+	// OIDs can be formatted as ".1.2.3..." ("absolute form") or "1.2.3..." ("relative form").
+	// Convert everything to relative form, like we do in the Python check.
+	return strings.TrimLeft(value, ".")
+}
+
+func parseSysUpTime(variable gosnmp.SnmpPDU) (uint32, error) {
+	name := NormalizeOID(variable.Name)
+	if name != sysUpTimeInstanceOID {
+		return 0, fmt.Errorf("expected OID %s, got %s", sysUpTimeInstanceOID, name)
+	}
+
+	value, ok := variable.Value.(uint32)
+	if !ok {
+		return 0, fmt.Errorf("expected uptime to be uint32 (got %v of type %T)", variable.Value, variable.Value)
+	}
+
+	return value, nil
+}
+
+func parseSnmpTrapOID(variable gosnmp.SnmpPDU) (string, error) {
+	name := NormalizeOID(variable.Name)
+	if name != snmpTrapOID {
+		return "", fmt.Errorf("expected OID %s, got %s", snmpTrapOID, name)
+	}
+
+	value := ""
+	switch variable.Value.(type) {
+	case string:
+		value = variable.Value.(string)
+	case []byte:
+		value = string(variable.Value.([]byte))
+	default:
+		return "", fmt.Errorf("expected snmpTrapOID to be a string (got %v of type %T)", variable.Value, variable.Value)
+	}
+
+	return NormalizeOID(value), nil
+}
+
+func parseVariables(variables []gosnmp.SnmpPDU) []trapVariable {
+	var parsedVariables []trapVariable
+
+	for _, variable := range variables {
+		varOID := NormalizeOID(variable.Name)
+		varType := formatType(variable)
+		varValue := formatValue(variable)
+		parsedVariables = append(parsedVariables, trapVariable{OID: varOID, VarType: varType, Value: varValue})
+	}
+
+	return parsedVariables
+}
+
+func formatType(variable gosnmp.SnmpPDU) string {
+	switch variable.Type {
+	case gosnmp.Integer, gosnmp.Uinteger32:
+		return "integer"
+	case gosnmp.OctetString:
+		return "string"
+	case gosnmp.ObjectIdentifier:
+		return "oid"
+	case gosnmp.Counter32:
+		return "counter32"
+	case gosnmp.Counter64:
+		return "counter64"
+	case gosnmp.Gauge32:
+		return "gauge32"
+	default:
+		return "other"
+	}
+}
+
+func formatValue(variable gosnmp.SnmpPDU) interface{} {
+	switch variable.Value.(type) {
+	case []byte:
+		return string(variable.Value.([]byte))
+	default:
+		return variable.Value
+	}
+}
+
+func formatVersion(packet *gosnmp.SnmpPacket) string {
+	switch packet.Version {
+	case gosnmp.Version3:
+		return "3"
+	case gosnmp.Version2c:
+		return "2"
+	case gosnmp.Version1:
+		return "1"
+	default:
+		return "unknown"
+	}
+}
```

Comment:
(no blocking for this PR since it was already there)

Do we need to report the SNMP version ?

Not sure if that's useful info, it's more related to the kind of credentials used (v2 community string, v3 user model) and not really part of the trap information. Also, a same device can support both v2 community string and v3 user model.

Reply:
Didn't want to change it in this PR. Do you think we should remove it?
The version doesn't provide I agree but it doesn't harm to have it

Reply:
Same as https://github.com/DataDog/datadog-agent/pull/10934#discussion_r825868959